### PR TITLE
API-2222: Swap tokens for requests coming from a user-restricted endpoint

### PR DESCRIPTION
**Pull request by gokyo**

### CODE REVIEW

File: app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala [hunk 1]
```diff
@@ -43,9 +43,11 @@ class UserRestrictedEndpointFilter @Inject()
...
       case Some(USER) => for {
         authority <- authorityService.findAuthority(proxyRequest)
-        isValidScope <- scopeValidator.validate(authority, requestHeader.tags.get(X_API_GATEWAY_SCOPE))
-      // TODO implement token swap
-      } yield requestHeader.withTag(X_APPLICATION_CLIENT_ID, authority.delegatedAuthority.clientId)
+        delegatedAuthority = authority.delegatedAuthority
+        isValidScope <- scopeValidator.validate(delegatedAuthority, requestHeader.tags.get(X_API_GATEWAY_SCOPE))
+      } yield requestHeader
+        .withTag(X_APPLICATION_CLIENT_ID, delegatedAuthority.clientId)
+        .withTag(AUTHORIZATION, delegatedAuthority.authBearerToken)
```
Comment (tom59) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: It should be withHeader, shouldnt it?
And it should be "Bearer ${authToken}"

Comment (edinhodzic) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: Just on the first point, there is no `withHeader` function. To add a "header" we would need to do `requestHeader.copy(headers = requestHeader.headers.add("foo" -> "bar"))`. The headers approach is more verbose so we've consciously stuck to using tags instead.

I may be mistaken but as far as I can tell headers are modelled as a wrappers around a sequence of tuples whereas tags are a raw map of strings to strings but effectively they serve the same function while tags have the added benefit of a more succinct api. There is probably a higher purpose to tags vs headers.

Agree with the second point.

File: app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala [hunk 2]
```diff
@@ -43,9 +43,11 @@ class UserRestrictedEndpointFilter @Inject()
     requestHeader.tags.get(X_API_GATEWAY_AUTH_TYPE) flatMap authType match {
       case Some(USER) => for {
         authority <- authorityService.findAuthority(proxyRequest)
-        isValidScope <- scopeValidator.validate(authority, requestHeader.tags.get(X_API_GATEWAY_SCOPE))
-      // TODO implement token swap
-      } yield requestHeader.withTag(X_APPLICATION_CLIENT_ID, authority.delegatedAuthority.clientId)
+        delegatedAuthority = authority.delegatedAuthority
+        isValidScope <- scopeValidator.validate(delegatedAuthority, requestHeader.tags.get(X_API_GATEWAY_SCOPE))
```
Comment (edgarjimenez) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: no need to name isValidScope, as it's not used anywhere.

Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: I know. 
It was ready for my next PR :)
```
_ <- scopeValidator.validate(...)
```

File: app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala [hunk 3]
```diff
@@ -43,9 +43,11 @@ class UserRestrictedEndpointFilter @Inject()
...
+      } yield requestHeader
```
Comment (edgarjimenez) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: I would consider moving this to updatedRequestHeader(delegatedAuthority...) for readability

File: app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala (hunk 1 above)
Comment (tom59) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: indeed just noticed you cant change the header of the requestHeader here. So I would suggest to have:
 .withTag(AUTHORIZATION, "Bearer ${delegatedAuthority.authBearerToken}")
And
      .withHeaders((AUTHORIZATION, request.tags(AUTHORIZATION))) in the ProxyConnector

File: app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala [hunk 4]
```diff
@@ -41,12 +41,18 @@ class UserRestrictedEndpointFilter @Inject()
...
-        authority <- authorityService.findAuthority(proxyRequest)
-        isValidScope <- scopeValidator.validate(authority, requestHeader.tags.get(X_API_GATEWAY_SCOPE))
-      // TODO implement token swap
-      } yield requestHeader.withTag(X_APPLICATION_CLIENT_ID, authority.delegatedAuthority.clientId)
+
+      case Some(USER) =>
+        for {
+          authority <- authorityService.findAuthority(proxyRequest)
+          delegatedAuthority = authority.delegatedAuthority
+          _ <- scopeValidator.validate(delegatedAuthority, requestHeader.tags.get(X_API_GATEWAY_SCOPE))
```
Comment (edinhodzic) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: This is a question just out of curiosity. Before we were passing an `Authority` instance to `scopeValidator`. Now we've gone into the structure of `Authority` and passing a `ThirdPartyDelegatedAuthority` instance instead  (`authority.delegatedAuthority`). What's the idea behind this?

I mean by that token we can equally well go into the structure of `ThirdPartyDelegatedAuthority` and pass a `Token` instance (`authority.delegatedAuthority.token`) or go further and pass a set of scopes (`authority.delegatedAuthority.token.scopes`).

Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: I thought to pass only the essential to `ScopeValidator`. 
We do not need the whole `Authority` object, so I decided to pass just `ThirdPartyDelegatedAuthority`.
However, we could just pass `authority.delegatedAuthority.token.scopes` (which is a `Set[String]`) as input parameter.
I guess it is a personal preference.
What do you suggest?

Comment (edinhodzic) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: That's the point I was trying to make when saying that the change may seem arbitrary i.e. what's _necessary_ is subjective therefore how deep we delve into the structure of `Authority` here, or whether we do it at all, is equally subjective. So when we say:

> We do not need the whole `Authority` object, so we decided to pass just `ThirdPartyDelegatedAuthority`.

We can now say:

> We do not need the whole `ThirdPartyDelegatedAuthority` object, so we decided to pass just the `Token`.

And then we can say:

> We do not need the whole `Token` object, so we decided to pass just the `Set[String]`.

And then we can go back to passing in an `Authority` instance and the self consuming cycle starts again.

So I'm not saying this change is wrong necessarily, just that the motivation behind the change is arbitrary/random ¯\\_(ツ)__/¯ hence my suggestion is that it was correct before but that it is also correct now with the caveat that we avoid going around in circles.

**Pull request merged**